### PR TITLE
Flows: Validation for flow keywords

### PR DESCRIPTION
**Pull request by AkhileshNegi**
## Summary
 Target issue is #2431

## Notes

- Added validation for Flows to check if a flow wait-for-response has keywords
- Added testcases
- Updated Test Workflow flow which is used in testcases

### CODE REVIEW

File: lib/glific/flows/flow.ex [hunk 1]
```diff
@@ -375,6 +375,7 @@ defmodule Glific.Flows.Flow do
         &Node.validate(&1, &2, flow)
       )
       |> dangling_nodes(flow)
+      |> check_flow_keywords(flow)
```
Comment (pankaj-ag) [inline lib/glific/flows/flow.ex]: @AkhileshNegi please take care of the consistency. We already have a validation structure in place. 
There should be a function called Wait.validate just like Node. validate and if it's not there please create one. 

Also, I think this validation should have come under Case.validate funcation.

File: lib/glific/flows/flow.ex [hunk 2]
```diff
@@ -411,6 +412,50 @@ defmodule Glific.Flows.Flow do
       else: [dangling: "Your flow has dangling nodes"] ++ errors
   end
 
+  @spec check_flow_keywords(Keyword.t(), map()) :: Keyword.t()
+  defp check_flow_keywords(errors, flow) do
+    all_nodes = flow_objects(flow, :node)
+
+    wait_for_response_words =
+      all_nodes
```
Comment (pankaj-ag) [inline lib/glific/flows/flow.ex]: This is a very expensive function we should clean it.

**Review by pankaj-ag — CHANGES_REQUESTED** [review]

File: lib/glific/flows/flow.ex (hunk 1 above)
Comment (AkhileshNegi) [reply lib/glific/flows/flow.ex]: @pankaj-ag Thanks for pointing out. I have restructured the code now

File: lib/glific/flows/flow.ex (hunk 2 above)
Comment (AkhileshNegi) [reply lib/glific/flows/flow.ex]: I have removed this and handling it in `Case.validate`

File: lib/glific/flows/case.ex [hunk 3]
```diff
@@ -84,8 +84,30 @@ defmodule Glific.Flows.Case do
   Validate a case
   """
   @spec validate(Case.t(), Keyword.t(), map()) :: Keyword.t()
-  def validate(_case, errors, _flow) do
-    errors
+  def validate(cases, errors, flow) do
```
Comment (pankaj-ag) [inline lib/glific/flows/case.ex]:
```suggestion
  def validate(case, errors, flow) do
```

File: lib/glific/flows/case.ex [hunk 4]
```diff
@@ -84,8 +84,30 @@ defmodule Glific.Flows.Case do
...
+    wait_for_response_words =
+      cases.arguments
+      |> List.first()
+      |> String.split(", ")
+      |> Enum.reduce(MapSet.new(), &MapSet.put(&2, &1))
```
Comment (pankaj-ag) [inline lib/glific/flows/case.ex]: I think we should clean the arguments as well with Glific clean.

File: lib/glific/flows/flow.ex [hunk 5]
```diff
@@ -366,6 +366,8 @@ defmodule Glific.Flows.Flow do
 
   @spec validate_flow(map()) :: Keyword.t()
   defp validate_flow(flow) do
+    all_nodes = flow_objects(flow, :node)
```
Comment (pankaj-ag) [inline lib/glific/flows/flow.ex]: Let's move this to else statement. We don't need to fetch all the nodes for empty flow.

File: lib/glific/flows/case.ex (hunk 3 above)
Comment (pankaj-ag) [inline lib/glific/flows/case.ex]:
```suggestion
  def validate(case, errors, flow) do
```

File: lib/glific/flows/case.ex [hunk 6]
```diff
@@ -84,8 +84,30 @@ defmodule Glific.Flows.Case do
...
+
+    if MapSet.disjoint?(flow_keywords, wait_for_response_words) do
+      errors
+    else
+      used_flow_keywords = MapSet.intersection(flow_keywords, wait_for_response_words)
+
+      Enum.reduce(
+        used_flow_keywords,
+        errors,
+        &(&2 ++ [flowContext: "\"#{&1}\" has already been used as a keyword for a flow"])
```
Comment (pankaj-ag) [inline lib/glific/flows/case.ex:108]: is there a possibility to add the flow name also in this message?

**Review by pankaj-ag — CHANGES_REQUESTED** [review]

File: lib/glific/flows/flow.ex (hunk 5 above)
Comment (AkhileshNegi) [reply lib/glific/flows/flow.ex]: @pankaj made requested changes

File: lib/glific/flows/case.ex (hunk 3 above)
Comment (AkhileshNegi) [reply lib/glific/flows/case.ex]: @pankaj made requested changes

Comment (AkhileshNegi) [reply lib/glific/flows/case.ex]: @pankaj made requested changes

File: lib/glific/flows/case.ex (hunk 6 above)
Comment (AkhileshNegi) [reply lib/glific/flows/case.ex:108]: @pankaj-ag dont see a use for that as flow can be searched using keyword as. will check with @abhi1203

Comment (abhi1203) [reply lib/glific/flows/case.ex:108]: Yeah that would be even better. if we can say `"Start" has already been used as a keyword for flow "flowname"` thanks for spotting

File: lib/glific/flows/case.ex (hunk 4 above)
Comment (AkhileshNegi) [reply lib/glific/flows/case.ex]: @pankaj-ag Thanks for discussion. I have updated the code now

**Review by pankaj-ag — APPROVED** [review]

**Pull request merged**